### ISSUE TITLE
[dart2js] Record `runtimeType` incorrectly affected by `runtimeType` overrides

### ISSUE THREAD

**Issue by askeksa-google** [dart-lang/sdk#51609]
Calling `runtimeType` on a record will return a record type with field types corresponding to the runtime types of the fields. In cases where a field is of a type that overrides `runtimeType`, the corresponding field type in the record type should reflect the actual runtime type of the field value, rather than what the overriding `runtimeType` implementation returns.

https://dart-review.googlesource.com/c/sdk/+/285684 (which fixes this for dart2wasm) contains a test case for this situation which fails on dart2js.

**Comment by fishythefish** [comment]
The expectation that `(a, b).runtimeType` isn't just `(a.runtimeType, b.runtimeType)` feels unintuitive to me. Is there a reason it has to use the actual runtime type? If a user finds it useful for `Foo().runtimeType` to be `Bar`, why isn't it useful for `(Foo(),).runtimeType` to be `(Bar,)`?

/cc @lrhn

**Comment by lrhn** [comment]
First of all `(a, b).runtimeType` can't be `(a.runtimeType, b.runtimeType)`, because the latter is a record object, not a `Type` object. I'm being pedantic because keeping the different categories separate is important here. 
We have types, `Type` objects, and other objects. Objects have types and `runtimeType` returns a `Type` Object. You can create a `Type` object from a type, but not the other way.

I expect `(a.runtimeType, b.runtimeType)` is intended to mean "a `Type` object representing the type of the pair with fields types `a.runtimeType` and `b.runtimeType`", but that's also not technically correct, because `a.runtimeType` is a `Type` object, not a type.
So, it'd have to be "a `Type` object representing the type of the pair with field types A and B, where A is the type represented by the `Type` object `a.runtimeType` and (similarly for B)".

That assumes that every `Type` object represents a type. That's not necessarily the case, because anyone can implement `Type`. We can't assume that the runtime system can go from an object of type `Type` to an actual type, which means using `a.runtimeType` to find the _type_ of `(a, b)` isn't necessarily well-defined.

Basically, we don't trust `runtimeType`, so we don't want to depend on it.

We're asking the *record* what its runtime type is, like if we're asking a `List<A>` about its runtime type, and we don't start asking the list elements for that.
You can think of a record as a generic class per shape, with type variables set to the *actual* type of each field value when the record is created. An `(1, 2)` is an `(int, int)` object, a `(_, _)`-shaped pair with `int` for each field.


To allow `a.runtimeType` to be used, we'd have to *keep* it as a `Type` object, and not worry about the result actually represents any valid type in the type system. That means, to create a record's `runtimeType` we'd have to:
* Call `.runtimeType` on every element, recursively if there are nested records.
* Create a `Type` object containing those other `Type` objects.
* Ensure that `Type` object has the correct equality and `hashCode` when compared to other `Type` objects.

It's not impossible. We'd have to introduce something like a new `Type` implementation:
```dart
class  _RecordType implements Type {
  final int _shapeId;
  final List<Type> _fieldRuntimeTypes;
  int get hashCode = Object.hash(_fieldRuntimeTypes) ^ _shapeId;
  bool operator==(Object other) {
    if (other is! _RecordType || _shapeId != other._shapeId) return false;
    for (var i = 0; i < _fieldRuntimeTypes.length; i++) {
      if (_fieldRuntimeTypes[i] != other._fieldRuntimeTypes[i]) return false;
    }
    return true;
  }
}
```
and use that as the `Type` object representation for _every_ `Record` type.

Then, if a field value lies on its `runtimeType`, that `Type` object (which can be an arbitrary used implemented object) will be asked on every `Type` equality check.

It's just more cumbersome than what we'd prefer. It requires creating `Type` objects for the field types, and not just storing the type internally, like we can do for generics.

**Comment by fishythefish** [comment]
Yes, I'm abusing notation for concision: in `(a, b).runtimeType`, I'm using `(_, _)` as a function `Object x Object -> Object`, and in `(a.runtimeType, b.runtimeType)`, I'm using it as a function `Type x Type -> Type` (implemented in practice as `_RecordType` or similar).

I don't think each `Type` here must correspond to an actual/honest type - as you say, we already don't trust `.runtimeType`. I'm more concerned with consistency between single values and tuples of those values when a user tries to debug with it. It's true that `_RecordType` carries more overhead than `.runtimeType` usually does, but does `.runtimeType` have to be an efficient operation?

**Comment by rakudrama** [comment]
Fixed by 1d4654c53068cf52f93c403ee0acee2a9e0caa23
